### PR TITLE
Add test cases for AST serialization

### PR DESCRIPTION
**Pull request by JukkaL**
The test cases use a serialized module in incremental mode
in various ways to make sure that all important information
in the AST nodes is preserved during serialization and
deserialization.

These are all very straightforward. The idea is to help catch
regressions early and to make it easier to refactor or optimize
serialization code. Working on these helped find 2 issues -- I
separated the first to PR #3275. The second issue (#3274) is 
still unfixed and there is a skipped test case for it.

More complex cases such as import cycles will still go to
check-incremental.test.

### CODE REVIEW

File: mypy/test/testcheck.py [hunk 1]
```diff
@@ -90,15 +91,22 @@ def cases(cls) -> List[DataDrivenTestCase]:
...
             # Incremental tests are run once with a cold cache, once with a warm cache.
             # Expect success on first run, errors from testcase.output (if any) on second run.
             # We briefly sleep to make sure file timestamps are distinct.
             self.clear_cache()
-            self.run_case_once(testcase, 1)
-            self.run_case_once(testcase, 2)
+            try:
+                self.run_case_once(testcase, 1)
+                self.run_case_once(testcase, 2)
+            finally:
```
Comment (gvanrossum) [inline mypy/test/testcheck.py]: There are now three try/finally blocks in this function that resture experiments.STRICT_OPTIONAL. Let's have a single one that does this, wrapping all different cases.

File: test-data/unit/check-serialize.test [hunk 2]
```diff
@@ -0,0 +1,1202 @@
+-- Serialization test cases (incremental type checking)
```
Comment (gvanrossum) [inline test-data/unit/check-serialize.test]: Where did you get all these tests? I hope you don't mind I'm not going to review all of them carefully, I trust you.

**Comment by JukkaL** [comment]
> Where did you get all these tests? I hope you don't mind I'm not going to review all of them carefully, I trust you.

I wrote down a list of interesting mypy features and manually created tests for those that seemed relevant. Yeah, it's okay not to review them carefully.

**Comment by gvanrossum** [comment]
Is this still blocked on quick mode crashes?

**Comment by JukkaL** [comment]
I've left out all test cases that cause quick mode crashes, so this can be merged. I'll merge this now since this only touches test cases.

**Pull request merged**